src/wazuh_modules/agent_upgrade/wm_agent_upgrade_command.c

```diff
@@ -0,0 +1,170 @@
+ * @param agents cJSON array with the agents_id
+ * @param task pointer to a task structure
+ * @param command command corresponding to the task
+ * @param response list of request to be sent to the task module. Expects cJSON array as input 
+ * @param failures list of request that failed to be added as tasks. Expects cJSON array as input 
+ * */
+static void wm_agent_create_upgrade_tasks(const cJSON *agents, wm_upgrade_task *task, const char* command, cJSON* response, cJSON* failures) {
+    assert(agents != NULL);
+    assert(task != NULL);
+    assert(command != NULL);
+    assert(response && (response->type == cJSON_Array));
+    assert(failures && (failures->type == cJSON_Array));
+
+    for(int i=0; i < cJSON_GetArraySize(agents); i++) {
+        wm_task *agent_task;
+        os_malloc(sizeof(wm_task), agent_task);
+        os_strdup(command, agent_task->command);
+        agent_task->task = task;
+        cJSON* agent_id = cJSON_GetArrayItem(agents, i);
+        int result = wm_agent_create_task_entry(agent_id->valueint, (void *)task);
+        if (result == OSHASH_SUCCESS ) {
+           cJSON *task_message = wm_agent_parse_task_module_message(agent_task->command, agent_id->valueint);
+           cJSON_AddItemToArray(response, task_message);
+        } else if (result == OSHASH_DUPLICATED) {
+            cJSON *task_message = wm_agent_parse_response_mesage(UPGRADE_ALREADY_ON_PROGRESS, upgrade_error_codes[UPGRADE_ALREADY_ON_PROGRESS], &(agent_id->valueint), NULL, NULL);
+            cJSON_AddItemToArray(failures, task_message);
+        } else {
+            cJSON *task_message = wm_agent_parse_response_mesage(UNKNOWN_ERROR, upgrade_error_codes[UNKNOWN_ERROR], &(agent_id->valueint), NULL, NULL);
+            cJSON_AddItemToArray(failures, task_message);
+        }
+    }
+}
+
+/**
+ * Sends json with task information to the task module and parses the response
+ * to give back to the api
+ * @param json_api cJSON array where the task response will be stored
+ * @param json_task_module cJSON to be sent to the task module 
+ * */
+static void wm_agent_parse_task_information(cJSON *json_api, const cJSON* json_task_module) {
+    cJSON *task_module_response = wm_agent_send_task_information(json_task_module);
+    if (task_module_response && (task_module_response->type == cJSON_Array)) {
+        // Parse task module responses into API
+        for(int i=0; i < cJSON_GetArraySize(task_module_response); i++) {
+            cJSON *task_response = cJSON_GetArrayItem(task_module_response, i);
+            int agent_id = cJSON_GetObjectItem(task_response, "agent")->valueint;
+            if (cJSON_HasObjectItem(task_response, "task_id")) {
+                // Store task_id
+                int task_id = cJSON_GetObjectItem(task_response, "task_id")->valueint;
+                wm_agent_insert_tasks_id(task_id, agent_id);
+                cJSON_AddItemReferenceToArray(json_api, task_response);
+            } else {
```

Comment:
This else will never happen since if there is an error in the task manager it will abort all the upgrades and return a JSON object with the error code, not an array.
Don't remove this code, maybe it is useful in the future.

Reply:
ok

Reply:
Never mind. Already changed this in tasks module and this code now is necessary.